### ISSUE TITLE
IllegalStateException: MQTT endpoint is closed

### ISSUE THREAD

**Issue by calohmn** [eclipse/hono#2030]
We saw the following exception in an MQTT adapter log:
````
ERROR io.vertx.core.impl.ContextImpl Unhandled exception
java.lang.IllegalStateException: MQTT endpoint is closed
	at io.vertx.mqtt.impl.MqttEndpointImpl.checkClosed(MqttEndpointImpl.java:724)
	at io.vertx.mqtt.impl.MqttEndpointImpl.close(MqttEndpointImpl.java:674)
	at io.vertx.mqtt.impl.MqttEndpointImpl.connack(MqttEndpointImpl.java:311)
	at io.vertx.mqtt.impl.MqttEndpointImpl.reject(MqttEndpointImpl.java:343)
	at io.vertx.mqtt.impl.MqttEndpointImpl.reject(MqttEndpointImpl.java:50)
	at org.eclipse.hono.adapter.mqtt.AbstractVertxBasedMqttProtocolAdapter.handleConnectionRequestResult(AbstractVertxBasedMqttProtocolAdapter.java:454)
	at org.eclipse.hono.adapter.mqtt.AbstractVertxBasedMqttProtocolAdapter.lambda$handleEndpointConnection$12(AbstractVertxBasedMqttProtocolAdapter.java:395)
	at io.vertx.core.impl.FutureImpl.tryFail(FutureImpl.java:197)
	at io.vertx.core.impl.FutureImpl.fail(FutureImpl.java:127)
[...]
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:152)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:113)
	at org.eclipse.hono.service.resourcelimits.PrometheusBasedResourceLimitChecks.lambda$executeQuery$24(PrometheusBasedResourceLimitChecks.java:494)
	at io.vertx.ext.web.client.impl.HttpContext.handleDispatchResponse(HttpContext.java:308)
	at io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:295)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:270)
	at io.vertx.ext.web.client.impl.predicate.PredicateInterceptor.handle(PredicateInterceptor.java:69)
	at io.vertx.ext.web.client.impl.predicate.PredicateInterceptor.handle(PredicateInterceptor.java:32)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:267)
	at io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:277)
	at io.vertx.ext.web.client.impl.HttpContext.dispatchResponse(HttpContext.java:238)
	at io.vertx.ext.web.client.impl.HttpContext.lambda$null$2(HttpContext.java:367)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:369)
[...]
````

There is no `CONNECT` tracing span in the OpenTracing component for the corresponding MQTT connection attempt, i.e. the exception causes the span not to get finished.

**Comment by sophokles73** [comment]
@calohmn which version are we talking about?

**Comment by calohmn** [comment]
@sophokles73 This occurred with Hono 1.2.3.

**Comment by calohmn** [comment]
This may happen in the following scenario:
The `verify message limit` check done as part of the CONNECT handling takes longer than the client is willing to wait for the CONACK, so that the client closes the MQTT connection. When the MQTT adapter is done with the message limit check and wants to reject the connection attempt (for whatever reason), this exception will occur.

**Comment by mbaeuerle** [comment]
We observed a similar behaviour when handling a high amount of connection attempts using Hono `1.2.1` but in this case the readiness probe for the HotrodCache was triggered before, which may have taken too long:
```
15:48:03.654 [vert.x-eventloop-thread-0] ERROR io.vertx.core.impl.ContextImpl - Unhandled exception
java.lang.IllegalStateException: MQTT endpoint is closed
    at io.vertx.mqtt.impl.MqttEndpointImpl.checkClosed(MqttEndpointImpl.java:724)
    at io.vertx.mqtt.impl.MqttEndpointImpl.close(MqttEndpointImpl.java:674)
    at io.vertx.mqtt.impl.MqttEndpointImpl.connack(MqttEndpointImpl.java:311)
    at io.vertx.mqtt.impl.MqttEndpointImpl.reject(MqttEndpointImpl.java:343)
    at io.vertx.mqtt.impl.MqttEndpointImpl.reject(MqttEndpointImpl.java:50)
    at org.eclipse.hono.adapter.mqtt.AbstractVertxBasedMqttProtocolAdapter.handleConnectionRequestResult(AbstractVertxBasedMqttProtocolAdapter.java:454)
    at org.eclipse.hono.adapter.mqtt.AbstractVertxBasedMqttProtocolAdapter.lambda$handleEndpointConnection$12(AbstractVertxBasedMqttProtocolAdapter.java:395)
    at io.vertx.core.impl.FutureImpl.tryFail(FutureImpl.java:197)
[...]
    at org.eclipse.hono.deviceconnection.infinispan.client.HotrodCache.lambda$checkForCacheAvailability$13(HotrodCache.java:377)
    at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:107)
[...]
    at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by calohmn** [comment]
@mbaeuerle It's not the readiness probe there but just the `HotrodCache.checkForCacheAvailability` method.
That method is invoked not only from the readiness probe but also from the `HotrodCache.isConnected` method. And that method gets called as part of the MQTT `CONNECT` handling (see `AbstractVertxBasedMqttProtocolAdapter.handleEndpointConnection` and the `isConnected` call in there).

So a possible scenario could have been that the Hotrod cache connection check (and the corresponding cache.put operation in there) took so long that the MQTT client already closed its connection attempt, which in the end lead to the above exception when the connection attempt was about to get rejected by the MQTT adapter.
The fix in #2057 will also cover that scenario.

That fact that doing a cache PUT on each device opening an MQTT connection seems unnecessary is addressed in #2068.

**Comment by calohmn** [comment]
Also fixed in 1.2.x (b94a67a4db43a514c93f50f9a1a66ef51b2db4cd)
